src/public/modules/forms/services/form-api.client.factory.js

```diff
@@ -25,129 +24,142 @@ const extractFormId = (path) => {
-        // all the form data at the top level. We need to ensure that the postprocessing
-        // is done in both cases.
-        if (get(response, 'resource.form.form_fields')) {
-          FormFields.injectMyInfoIntoForm(response.resource.form)
-          // Convert plain form object to smart Form instance
-          response.resource.form = new Form(response.resource.form)
-        } else if (get(response, 'resource.form_fields')) {
-          FormFields.injectMyInfoIntoForm(response.resource)
-          response.resource = new Form(response.resource)
-        }
-        return response.resource
-      },
+  const handleResponse = (data) => {
+    // The backend returns different shapes for different request types. For GET
+    // requests, it returns an object with a form attribute containing
+    // all the form data; for PUT requests, it returns an object with
+    // all the form data at the top level. We need to ensure that the postprocessing
+    // is done in both cases.
+    if (get(data, 'form.form_fields')) {
+      FormFields.injectMyInfoIntoForm(data.form)
+      // Convert plain form object to smart Form instance
+      data.form = new Form(data.form)
+    } else if (get(data, 'form_fields')) {
+      FormFields.injectMyInfoIntoForm(data)
+      data = new Form(data)
     }
+    return data
+  }
+
+  /**
+   * Handles error returned from API calls. In charge of redirecting users using FormErrorService
+   * if redirectOnError is true.
+   * @param {Error} err Error returned from API calls
+   * @param {{redirectOnError: boolean, errorTargetState: string }} errorParams redirectOnError specifies
+   * whether to use FormErrorService to redirect to errorTargetState when an error is encountered.
+   * If redirectOnError is true, this is the redirect state which will be passed to FormErrorService
+   */
+  const handleError = (err, errorParams) => {
+    const { redirectOnError, errorTargetState } = errorParams
     if (redirectOnError) {
-      interceptor.responseError = (response) => {
-        return FormErrorService.redirect({
-          response,
-          targetState: errorTargetState,
-          targetFormId: extractFormId(get(response, 'config.url')),
-        })
-      }
+      FormErrorService.redirect({
+        response: err.response,
+        targetState: errorTargetState,
+        targetFormId: extractFormId(get(err.response, 'config.url')),
+      })
+    } else {
+      throw err // just pass error on
     }
-    return interceptor
   }
 
-  // accessMode is either adminForm or publicForm
-  let resourceUrl = '/:formId/:accessMode'
-  const V3_PUBLICFORM_URL = '/api/v3/forms/:formId'
+  const generateService = (service, errorParams, ...inputs) => {
+    return service(...inputs.map((input) => handleInput(input)))
+      .then((data) => {
+        return handleResponse(data)
+      })
+      .catch((err) => handleError(err, errorParams))
+  }
```

Comment:
while I see how this reduces repeated code, I would actually rather call the Myinfo injection/stripping code explicitly in each function and only when it is relevant. this would be more verbose, but would make the pipeline for each API call very explicit according to the shape. rn it's hard for me to tell what transformations are being applied to the data for each API call, and this might also break silently if e.g. we change the param for a service from `{ form: someObject }` to just `someObject`. wdyt?

Reply:
Oh yes thats true. Initially I rewrote it as such because this actually mimics the behaviour of the previous request and response interceptors when we were using the `$resource` object. I think it does makes sense to call the injection/stripping code explicitly now since we are refactoring. However, would it be alright if I keep this function and just pass in  the functions `handleInput` and `handleResponse` so that it is specific to each API call? I personally think that this `generateService` function keeps the code neater.

Reply:
sure that sounds good